Refactors cleaning

kawoppi opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
A cleaner component has been added so that the cleaning code can be put in one place. This cleaner component has been used to replace a lot of the cleaning code of mops/soap/rags/cleanbots. The cleaner component can handle cleaning while accounting for cleaning skill. Any cases where simply washing the target doesn't suffice can also be handled here. Such as washing bloodied windows for example. The cleaner component also has a callback it can call after successfully cleaning something. Soap makes use of this to decrease the amount of uses it has left. Mops make use of this to place reagents on the cleaned floor and reduce the reagents it has left.

Rags are now more in line with soap, because they both use the cleaner component. This means rags can get increased cleaning speed based on your cleaning skill. You will also be getting cleaning experience when you clean with rags from now on.

Mops and cleanbots still only clean turfs. The cleaner component doesn't need to care about this detail though. They just pass the turf to it as a parameter.

Soap used to not be able to clean the clothing you were wearing. According to a code comment this was because it caused overlay bugs. Cleaning worn items seems to be updating the sprites just fine nowadays though, so this restriction has been removed.

There were some code comments stating that mops were supposed to get less experience than soap. This was supposed to be done through rounding the experience gain for some reason. But the experience gain for soap and mops seemed to be the same. These comments are now gone. There's an experience gain modifier on the cleaner datum now though, in case somebody wants to do something like that in the future. And it will be much more readable that way too.


Sorry for the amount of commits, I made the mistake of not starting from master. Relevant commits are from [added basic cleaner datum, applied it to rags](https://github.com/tgstation/tgstation/pull/68156/commits/a8eb5feeee6312f0a50c693019be465b3b404825) and onwards.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->


## Why It's Good For The Game
Having similar code doing similar things in four different places makes it harder to expand that functionality. It also causes little inconsistencies such as the rags not giving cleaning experience for example. Having the cleaning code in one place makes it easier to expand and fix.

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->


## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
fix: rag cleaning speed is now influenced by your cleaning skill
fix: cleaning with a rag now gives you cleaning skill experience
fix: rags can now be used to wash bloodied windows
fix: soap is no longer restricted from washing worn clothing, as this doesn't seem to be causing issues anymore
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

kawoppi committed:
fuck

clicking again just removes the overlay

kawoppi committed:
initial attempt

clicking again just removes the overlay :(

kawoppi committed:
damp rag overlay now functions correctly

thanks friedpotato

kawoppi committed:
damp rag overlay now functions correctly

thanks friedpotato

kawoppi committed:
applied fix to mop and soap

kawoppi committed:
applied fix to mop and soap

kawoppi committed:
added sloshing animation

kawoppi committed:
added sloshing animation

kawoppi committed:
changed to bubble animation

blubblubblubblub

kawoppi committed:
changed to bubble animation

blubblubblubblub

kawoppi committed:
changed overlay layer

it needs to appear above the filth

kawoppi committed:
changed overlay layer

it needs to appear above the filth

kawoppi committed:
added comments and adjusted spacing

Felt like that would make it more readable, though maybe more experienced people could read it just fine before.

kawoppi committed:
added comments and adjusted spacing

Felt like that would make it more readable, though maybe more experienced people could read it just fine before.

kawoppi committed:
changed cleaning overlay layer

You can now see it over tables, gibs and more.

kawoppi committed:
changed cleaning overlay layer

You can now see it over tables, gibs and more.

kawoppi committed:
Merge remote-tracking branch 'origin/dev/cleaning-animation' into dev/cleaning-animation

kawoppi committed:
Merge branch 'master' into dev/cleaning-animation

kawoppi committed:
cleaning under people with a rag now looks better

Bubbles display under people normally. If you're directly cleaning a person or anything else with a higher plane than the lower overlay it will use the higher overlay.

It doesn't display above potted plants though, but at least it's hidden outside of FOV.

kawoppi committed:
no more FOV hidden, display it under furniture instead

I'd rather have both but they are different planes. Seeing blood and gibs suddenly disappear outside of your FOV sorta has the same effect so I guess it's not a disaster.

kawoppi committed:
improved overlay picking code

Hopefully this is more readable. I think it also prevents a potential issue when what you're cleaning is on a lower plane but it's layer is higher.

This will be quite some code to copy, I really oughta give components another try sometime.

kawoppi committed:
update mops and soap to rag standards

kawoppi committed:
added cleaning animation to cleanbot

That makes four places I've copied this code to... Shit.

kawoppi committed:
added basic cleaner datum, applied it to rags

kawoppi committed:
allows the cleaning duration to be modified

kawoppi committed:
skill modifier can now be offset

This is done for soap. Took me a while to realize that the "speedies" getting reduced actually meant the duration getting shorter. That's why I'm calling it a duration in here now.

kawoppi committed:
cleaning strength can now be set

kawoppi committed:
improved code comments

kawoppi committed:
added documentation to clean proc

kawoppi committed:
cleaning gives experience

The mop and soap get EXP differently... A comment in the mop code reads: "it is intentional that the mop rounds xp but soap does not, USE THE SACRED TOOL". I have opted to add a skill modifier instead because like what the fuck man.

kawoppi committed:
generalized cleaning message

kawoppi committed:
carried over extra soap functionality

Carried over the extra cleaning functionality that soap had. This was washing people's mouths and washing windows.

Cleaning items while wearing them no longer seems to be an issue, so that has not been carried over from soap.

kawoppi committed:
made two nullchecks smaller

kawoppi committed:
added datum and constructor code comments

kawoppi committed:
moved animation overlays to the cleaner file

kawoppi committed:
rag cleaner uses old rag duration

Better put this here explicitly before I might change the default and forget about it.

kawoppi committed:
soap now makes use of the cleaner

kawoppi committed:
callbacks get both parameters

kawoppi committed:
updated soap proc comment

kawoppi committed:
moved mouth washing code back to soap

It doesn't make as much sense when you do it using something that isn't soap.

kawoppi committed:
re-added soap speed scaling

Also renamed the offset name, because I got confused by my own name. Speed was a bad name because if you increase it you'd think the duration would be shorter.

kawoppi committed:
cleanbot now makes use of the cleaner datum

The initial cleanbot cleaning duration was just 1. You can barely see the bot animation, but I think I remember you used to be able to see it for longer. I'll keep it as it was for now.

kawoppi committed:
fixed cleanbot cleaning duration

Seems like the duration of 1 was an error. Somebody probably forgot to use the SECONDS define.

kawoppi committed:
mop now makes use of the cleaner datum

kawoppi committed:
renamed mop.clean() proc to apply_reagents()

kawoppi committed:
removed clean_start_callback

Yeah you can just do the checks before calling clean() so the callback is unnecessary.

kawoppi committed:
removed cleaning animation

It was here because I started from my cleaning animation branch. I'll bring it back if the PR for it gets merged.

kawoppi committed:
are you happy now, mister linter?

kawoppi committed:
cleaners now get deleted

Not sure if I'm doing this right.

kawoppi commented on code/datums/cleaner.dm:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Can be used to clean things.
+ * Takes care of duration, cleaning skill and special cleaning interactions.
+ * A callback can be set by the datum holding the cleaner to add custom functionality.
+ * Soap uses a callback to decrease the amount of uses it has left after cleaning for example.
+ */
+/datum/cleaner
```
Originally I tried to make it a component, but I had some trouble trying with using the afterattack signal. Though now I realize I probably could've just made a new signal to use for cleaning when the afterattack doesn't return early. I think I'll give that another go.

cacogen commented:
I remember I sped up cleanbot cleaning speed a few years ago in an effort to make them less useless.

It was probably before the SECONDS define, but I think it was intentional that it clean in 100ms because even with that immediacy they still take forever to path to mess and bunch up when using more than one of them, rendering them fairly useless.

I don't think you should nerf them  when they're already pretty slow just to see more of the animation.

kawoppi commented:
Thanks for letting me know. I thought that change was made in error so I'll revert it back to 0.1 seconds.

kawoppi committed:
reverted cleanbot cleaning duration to 0.1 seconds

kawoppi committed:
added cleaner component, used it for rags

kawoppi committed:
added cleaning signal define

kawoppi committed:
updated code comments

kawoppi committed:
added cleaner component to cleanbot

kawoppi committed:
added cleaner component to soap

kawoppi committed:
added cleaner component to mop

kawoppi committed:
fixed signal handler

I thought it might be better to keep all of the cleaning behavior under the same proc instead of splitting it up.

kawoppi committed:
removed outdated cleaner datum

kawoppi committed:
added source parameter to callback

I'm not using it right now, but someone might probably have a use for it in the future.

kawoppi committed:
fixed soap oversight, whoopsie

kawoppi committed:
added destructor to cleaner component

kawoppi committed:
forgot to call the parent

JohnFulpWillard commented on code/datums/components/cleaner.dm:
```diff
@@ -0,0 +1,88 @@
+	/// Gets called when something is successfully cleaned.
+	var/datum/callback/on_cleaned_callback
+	/// Gets invoked asynchronously by the signal handler
+	var/datum/callback/cleaning_proc
```
Do we need this var or can we just ``INVOKE_ASYNC(src, .proc/clean, source, target, user)`` when needed?

JohnFulpWillard commented on code/game/objects/items/mop.dm:
```diff
@@ -29,24 +29,25 @@
-		A.wash(CLEAN_SCRUB)
-
-	reagents.expose(A, TOUCH, 10) //Needed for proper floor wetting.
+/obj/item/mop/ComponentInitialize()
```
Can you use Initialize() instead of ComponentInitialize() ? It's deprecated and we would like to remove it in the future.

JohnFulpWillard commented on code/game/objects/items/mop.dm:
```diff
@@ -62,13 +63,13 @@
-			to_chat(user, span_notice("You finish mopping."))
-			clean(T, user)
+		var/datum/component/cleaner/cleaner = GetComponent(/datum/component/cleaner)
+		if(!reagents.has_chemical_flag(REAGENT_CLEANS, 1)) //won't clean the turf nor give you experience without cleaning reagents
```
Can't we just early return if there's no chemicals, rather than GetComponent and set to not give XP?

kawoppi replied:
Can't early return because you can still mop with chemicals that don't actually clean. So if you dip your mop in welding fuel for example, you still get the do_after, skill speed modifier, the cleaning message and it applies the welding fuel to the floor. Setting the washing strength and XP modifier was an easy way to cheat and use the same code without actually washing the tile. I can't really think of a way to avoid using GetComponent without having to duplicate some code that the cleaning component is already supposed to take care of.

JohnFulpWillard replied:
You can make it an arg for COMSIG_START_CLEANING, maybe. It would be preferred over GetComponent and setting exp gain to 0, cleaning, then setting it back to 1.

JohnFulpWillard reviewed (commented):
sorry for not putting it all in one review

kawoppi committed:
replaced callback with async call

kawoppi committed:
removed usage of ComponentInitialize, moved components to Initialize

kawoppi committed:
applied suggestions from code review

hell yeah

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>

kawoppi committed:
empty commit

kawoppi committed:
removed usage of GetComponent()

I know I might be doubling up on defaulting clean_target to true, but having the clean proc not clean by default would be confusing.

kawoppi committed:
added clean_target to callback parameters

Someone might need this in the future

kawoppi committed:
I keep forgetting to omit var/ from parameters

JohnFulpWillard commented on code/datums/components/cleaner.dm:
```diff
@@ -0,0 +1,93 @@
+ * * user the person doing the cleaning
+ * * clean_target set this to false if the target should not be washed and if experience should not be awarded to the user
+ */
+/datum/component/cleaner/proc/clean(datum/source, atom/target as obj|turf|area, mob/living/user, clean_target = TRUE)
```
```suggestion
/datum/component/cleaner/proc/clean(datum/source, atom/target, mob/living/user, clean_target = TRUE)
```

JohnFulpWillard commented on code/datums/components/cleaner.dm:
```diff
@@ -0,0 +1,93 @@
+
+	//do the cleaning
+	user.visible_message(span_notice("[user] starts to clean [target]!"), span_notice("You start to clean [target]..."))
+	if(do_after(user, cleaning_duration, target = target))
```
You think we can do an early return here with if(!do_after()) instead?

JohnFulpWillard commented on code/game/objects/items/mop.dm:
```diff
@@ -62,13 +60,10 @@
-		if(do_after(user, mopspeed*clean_speedies, target = T))
-			to_chat(user, span_notice("You finish mopping."))
-			clean(T, user)
+		if(reagents.has_chemical_flag(REAGENT_CLEANS, 1))
+			SEND_SIGNAL(src, COMSIG_START_CLEANING, T, user)
+		else //won't clean the turf nor give you experience without cleaning reagents
+			SEND_SIGNAL(src, COMSIG_START_CLEANING, T, user, FALSE)
```
instead of sending 2 signals, and since has_chemical_flag() only returns TRUE/FALSE, can't we do something instead like
```suggestion
		var/should_clean = reagents.has_chemical_flag(REAGENT_CLEANS, 1)
		SEND_SIGNAL(src, COMSIG_START_CLEANING, T, user, clean_target = should_clean)
```

JohnFulpWillard reviewed (commented):
these should be my last problems, sorry for taking so much of your time.

kawoppi committed:
Apply suggestions from code review

Co-authored-by: John Willard <53777086+JohnFulpWillard@users.noreply.github.com>

kawoppi commented:
ah don't worry about it, thanks for the code reviews!

kawoppi committed:
do_after early returns if false

kawoppi committed:
added default cleaning duration

kawoppi committed:
removed experience_gain_modifier

kawoppi committed:
used atom level proc for sending signals

kawoppi committed:
atom level not datum level

kawoppi commented on code/modules/reagents/reagent_containers/misc.dm:
```diff
@@ -152,7 +156,4 @@
-		if(do_after(user,30, target = A))
-			user.visible_message(span_notice("[user] finishes wiping off [A]!"), span_notice("You finish wiping off [A]."))
-			A.wash(CLEAN_SCRUB)
+		SEND_SIGNAL(src, COMSIG_START_CLEANING, A, user)
```
I've added a proc to the cleaner component file. Is that the right place for it?

JohnFulpWillard replied:
Should put it in ``code/game/atoms.dm`` instead

kawoppi committed:
moved cleaning proc to atoms.dm
